hive/agents/ppo.py

```diff
@@ -0,0 +1,411 @@
+                agent should use to update itself.
+        """
+        if self._reward_clip is not None:
+            update_info["reward"] = np.clip(
+                update_info["reward"], -self._reward_clip, self._reward_clip
+            )
+        preprocessed_update_info = {
+            "observation": update_info["observation"],
+            "action": self.scale_action(update_info["action"]),
+            "reward": update_info["reward"],
+            "done": update_info["done"],
+        }
+        if "agent_id" in update_info:
+            preprocessed_update_info["agent_id"] = int(update_info["agent_id"])
+
+        return preprocessed_update_info
+
+    def preprocess_update_batch(self, batch):
+        """Preprocess the batch sampled from the replay buffer.
+
+        Args:
+            batch: Batch sampled from the replay buffer for the current update.
+
+        Returns:
+                - Preprocessed batch.
+        """
+        for key in batch:
+            batch[key] = torch.tensor(batch[key], device=self._device)
+        return batch
+
+    @torch.no_grad()
+    def act(self, observation):
+        """Returns the action for the agent. 
+
+        Args:
+            observation: The current observation.
+        """
+
+        observation = torch.tensor(
+            np.expand_dims(observation, axis=0), device=self._device
+        ).float()
+        #TODO: can add random exploration
+        action, logprob, _ = self._actor(observation)
+        value = self._critic(observation)
+        action = action.cpu().numpy()
+        if self._continuous_action:
+            action = self.unscale_actions(action)
+            action = np.clip(action, self._action_min, self._action_max)
+
+        #TODO: remove if vectorized environments
+        action = np.squeeze(action)
+
+        self._logprob = logprob.cpu().numpy()
+        self._value = value.cpu().numpy()
+        return action
+
+    def update(self, update_info):
+        """
+        Updates the PPO agent.
+
+        Args:
+            update_info: dictionary containing all the necessary information to
+                update the agent. Should contain a full transition, with keys for
+                "observation", "action", "reward", and "done".
+        """
+        if not self._training:
+            return
+        
+        # Inspired  by https://github.com/vwxyzjn/ppo-implementation-details/blob/main/ppo_shared.py
+        if (self._replay_buffer.ready()):
+            self._replay_buffer.compute_advantages(self._value)
+            #TODO: recheck sampling
+            for _ in range(self._num_epochs):
+                valid_ind_size = self._replay_buffer._find_valid_indices()
+                for _ in range(valid_ind_size//self._batch_size):
+                    batch = self._replay_buffer.sample(batch_size=self._batch_size)
+                    batch = self.preprocess_update_batch(batch)
+                    self._actor_optimizer.zero_grad()
```

Comment:
If we dont have one shared architecture, we have to deal with repeating actor and critic commands everytime